### ISSUE TITLE
Auto property from constructor parameter refactoring provider errors

### ISSUE THREAD

**Issue by mrward** [mono/monodevelop#5519]
VS Mac 7.6.0.2093

Release ID: 706002093
Git revision: e6a5c636b10e72f8ca3266684dae54de298a6cd6
Build date: 2018-07-20 19:42:19+00
Build branch: release-7.6
Xamarin extensions: 6b3acb336272faaa9c7e0f76597a33fa7fd31819

Typing in a constructor shows two status bar errors are displayed for InitializeAutoPropertyFromConstructorParameterCodeRefactoringProvider and InitializeReadOnlyAutoPropertyFromConstructorParameterCodeRefactoringProvider. Seems to happen if the parameter name starts with C# keyword, such as 'for' or 'if'.

![roslynstatusbarerrorctor](https://user-images.githubusercontent.com/372361/43256213-fa782a10-90c3-11e8-88d7-30b95e29accb.gif)

IDE logs have index out of range errors:

```
ERROR [2018-07-26 11:01:58Z]: Roslyn error: Extension_Exception InitializeAutoPropertyFromConstructorParameterCodeRefactoringProvider : Index was outside the bounds of the array.
  at RefactoringEssentials.CSharp.ConstructorParameterContextFinder.GetPropertyName (System.String v) [0x00000] in /Users/vsts/agent/2.136.1/work/1/s/monodevelop/main/external/RefactoringEssentials/RefactoringEssentials/CSharp/ConstructorParameterContextFinder.cs:44 
  at RefactoringEssentials.CSharp.ConstructorParameterContextFinder+<Find>d__0.MoveNext () [0x001e1] in /Users/vsts/agent/2.136.1/work/1/s/monodevelop/main/external/RefactoringEssentials/RefactoringEssentials/CSharp/ConstructorParameterContextFinder.cs:39 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:357 
  at RefactoringEssentials.CSharp.CodeRefactorings.InitializeAutoPropertyFromConstructorParameterCodeRefactoringProvider+<ComputeRefactoringsAsync>d__4.MoveNext () [0x00033] in /Users/vsts/agent/2.136.1/work/1/s/monodevelop/main/external/RefactoringEssentials/RefactoringEssentials/CSharp/CodeRefactorings/Synced/InitializeAutoPropertyFromConstructorParameterCodeRefactoringProvider.cs:25 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:447 
  at Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService+<GetRefactoringFromProviderAsync>d__8.MoveNext () [0x000a0] in /_/src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs:143 
ERROR [2018-07-26 11:02:30Z]:ERROR [2018-07-26 11:02:30Z]: Roslyn error: Extension_Exception InitializeReadOnlyAutoPropertyFromConstructorParameterCodeRefactoringProvider : Index was outside the bounds of the array.
  at RefactoringEssentials.CSharp.ConstructorParameterContextFinder.GetPropertyName (System.String v) [0x00000] in /Users/vsts/agent/2.136.1/work/1/s/monodevelop/main/external/RefactoringEssentials/RefactoringEssentials/CSharp/ConstructorParameterContextFinder.cs:44 
  at RefactoringEssentials.CSharp.ConstructorParameterContextFinder+<Find>d__0.MoveNext () [0x001e1] in /Users/vsts/agent/2.136.1/work/1/s/monodevelop/main/external/RefactoringEssentials/RefactoringEssentials/CSharp/ConstructorParameterContextFinder.cs:39 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:357 
  at RefactoringEssentials.CSharp.CodeRefactorings.InitializeReadOnlyAutoPropertyFromConstructorParameterCodeRefactoringProvider+<ComputeRefactoringsAsync>d__4.MoveNext () [0x00033] in /Users/vsts/agent/2.136.1/work/1/s/monodevelop/main/external/RefactoringEssentials/RefactoringEssentials/CSharp/CodeRefactorings/Custom/InitializeReadOnlyAutoPropertyFromConstructorParameterCodeRefactoringProvider.cs:21 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:447 
  at Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService+<GetRefactoringFromProviderAsync>d__8.MoveNext () [0x000a0] in /_/src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs:143 
 Roslyn error: Extension_Exception InitializeAutoPropertyFromConstructorParameterCodeRefactoringProvider : Index was outside the bounds of the array.
  at RefactoringEssentials.CSharp.ConstructorParameterContextFinder.GetPropertyName (System.String v) [0x00000] in /Users/vsts/agent/2.136.1/work/1/s/monodevelop/main/external/RefactoringEssentials/RefactoringEssentials/CSharp/ConstructorParameterContextFinder.cs:44 
  at RefactoringEssentials.CSharp.ConstructorParameterContextFinder+<Find>d__0.MoveNext () [0x001e1] in /Users/vsts/agent/2.136.1/work/1/s/monodevelop/main/external/RefactoringEssentials/RefactoringEssentials/CSharp/ConstructorParameterContextFinder.cs:39 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:357 
  at RefactoringEssentials.CSharp.CodeRefactorings.InitializeAutoPropertyFromConstructorParameterCodeRefactoringProvider+<ComputeRefactoringsAsync>d__4.MoveNext () [0x00033] in /Users/vsts/agent/2.136.1/work/1/s/monodevelop/main/external/RefactoringEssentials/RefactoringEssentials/CSharp/CodeRefactorings/Synced/InitializeAutoPropertyFromConstructorParameterCodeRefactoringProvider.cs:25 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:447 
  at Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService+<GetRefactoringFromProviderAsync>d__8.MoveNext () [0x000a0] in /_/src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs:143 
```

> VS bug [#654072](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/654072)

**Comment by mrward** [comment]
@monojenkins backport release-7.6

**Comment by xamarin-release-manager** [comment]
Fixed in version 7.7.0.766 (master)

Author: Mike Krüger
Commit: 2dc0c1e8ab5606c7d3b7a59f6987b47d9ea9f3a1 (mono/monodevelop)

**Comment by parasAmbhore** [comment]
I can confirm that this issue is not reproducible on below build.
Hence marking this as Fix Tested.

Build Info:

=== Visual Studio Enterprise 2017 for Mac (Preview) ===
Version 7.7 Preview (7.7 build 762)
Mono 5.10.1.57 (2017-12/ea8a24b1bbf) (64-bit)
Xamarin.Mac 4.4.1.178 (master / eeaeb7e6)

=== Xamarin.Mac ===
Version: 4.6.0.11 (Visual Studio Enterprise)
=== Xamarin.iOS ===
Version: 11.14.0.10 (Visual Studio Enterprise)
=== Xamarin.Android ===
Version: 9.0.0.17 (Visual Studio Enterprise)

Build Taken from: https://dl.internalx.com/artifacts/devdiv-1895915/VisualStudioForMac-Preview-7.7.0.762.dmg
Screen Cast: https://www.screencast.com/t/HRfA00B20X
Gist Link: https://gist.github.com/mohakbarokar/a67d38ad8d1f4c4e96812f05907f96f8

**Comment by xamarin-release-manager** [comment]
Fixed in version 7.6.0.2113 (release-7.6)

Author: Mike Krüger
Commit: b7b94a1518a5c64a83c62bdbe89a0d62c895cc94 (mono/monodevelop)
